frame/contracts/src/wasm/mod.rs

```diff
@@ -425,27 +426,51 @@ mod tests {
 			self.terminations.push(TerminationEntry { beneficiary: beneficiary.clone() });
 			Ok(())
 		}
-		fn get_storage(&mut self, key: &StorageKey) -> Option<Vec<u8>> {
-			self.storage.get(key).cloned()
+		fn get_storage(&mut self, key: &FixSizedKey) -> Option<Vec<u8>> {
+			self.storage.get(&key.to_vec()).cloned()
 		}
-		fn get_storage_size(&mut self, key: &StorageKey) -> Option<u32> {
-			self.storage.get(key).map(|val| val.len() as u32)
+		fn get_storage_transparent(&mut self, key: &VarSizedKey<Self::T>) -> Option<Vec<u8>> {
+			self.storage.get(&key.to_vec()).cloned()
+		}
+		fn get_storage_size(&mut self, key: &FixSizedKey) -> Option<u32> {
+			self.storage.get(&key.to_vec()).map(|val| val.len() as u32)
+		}
+		fn get_storage_size_transparent(&mut self, key: &VarSizedKey<Self::T>) -> Option<u32> {
+			self.storage.get(&key.to_vec()).map(|val| val.len() as u32)
 		}
 		fn set_storage(
 			&mut self,
-			key: StorageKey,
+			key: &FixSizedKey,
 			value: Option<Vec<u8>>,
 			take_old: bool,
 		) -> Result<WriteOutcome, DispatchError> {
-			let entry = self.storage.entry(key);
+			let entry = self.storage.entry(key.to_vec());
 			let result = match (entry, take_old) {
 				(Entry::Vacant(_), _) => WriteOutcome::New,
 				(Entry::Occupied(entry), false) =>
 					WriteOutcome::Overwritten(entry.remove().len() as u32),
 				(Entry::Occupied(entry), true) => WriteOutcome::Taken(entry.remove()),
 			};
 			if let Some(value) = value {
-				self.storage.insert(key, value);
+				self.storage.insert(key.to_vec(), value);
+			}
+			Ok(result)
+		}
+		fn set_storage_transparent(
+			&mut self,
+			key: &VarSizedKey<Self::T>,
+			value: Option<Vec<u8>>,
+			take_old: bool,
+		) -> Result<WriteOutcome, DispatchError> {
+			let entry = self.storage.entry(key.to_vec());
+			let result = match (entry, take_old) {
+				(Entry::Vacant(_), _) => WriteOutcome::New,
+				(Entry::Occupied(entry), false) =>
+					WriteOutcome::Overwritten(entry.remove().len() as u32),
+				(Entry::Occupied(entry), true) => WriteOutcome::Taken(entry.remove()),
+			};
+			if let Some(value) = value {
+				self.storage.insert(key.to_vec(), value);
 			}
```

Comment:
We should de duplicate this code while we can't have these functions itself generic: Create a generic function in this module which you call from both of these function.

Reply:
What signature do you mean for that function? if its like this 
```rust
	fn set_storage_generic<K: StorageKey<T>>(
		&mut self,
		key: &K,
		value: Option<Vec<u8>>,
		take_old: bool,
	) -> Result<WriteOutcome, DispatchError> {

```
then this won't compile with 
```
`exec::Ext` cannot be made into an object
```

This could be made like this
```rust
	fn set_storage_generic(
		&mut self,
		key: Vec<u8>,
		value: Option<Vec<u8>>,
		take_old: bool,
	) -> Result<WriteOutcome, DispatchError> {
```

but once `storage::write()` function [accepts](https://github.com/paritytech/substrate/blob/ag-transparent-hashing/frame/contracts/src/storage.rs#L146) generic parameter, this won't make much sense either

Reply:
I didn't realize that this is only the `MockExt`. This is fine then.